File: app/src/main/java/net/squanchy/tweets/TwitterComponent.java
package net.squanchy.tweets;

import net.squanchy.injection.ActivityLifecycle;
import net.squanchy.injection.ApplicationComponent;
import net.squanchy.tweets.service.TwitterRepository;
import net.squanchy.tweets.service.TwitterService;

import dagger.Component;

@ActivityLifecycle
@Component(modules = TwitterModule.class, dependencies = ApplicationComponent.class)
interface TwitterComponent {

    TwitterRepository repository();

    TwitterService service();
}

File: app/src/main/java/net/squanchy/tweets/TwitterComponent.kt
package net.squanchy.tweets

import androidx.appcompat.app.AppCompatActivity
import dagger.Component
import net.squanchy.injection.BaseActivityComponentBuilder
import net.squanchy.injection.ActivityLifecycle
import net.squanchy.injection.ApplicationComponent
import net.squanchy.injection.applicationComponent
import net.squanchy.navigation.NavigationModule
import net.squanchy.navigation.Navigator
import net.squanchy.tweets.service.TwitterService

internal fun twitterComponent(activity: AppCompatActivity): TwitterComponent =
    DaggerTwitterComponent.builder()
        .applicationComponent(activity.applicationComponent)
        .activity(activity)
        .build()

@ActivityLifecycle
@Component(
    modules = [(TwitterModule::class), (NavigationModule::class)],
    dependencies = [(ApplicationComponent::class)]
)
internal interface TwitterComponent {

    fun service(): TwitterService

    fun navigator(): Navigator

    @Component.Builder
    interface Builder : BaseActivityComponentBuilder<TwitterComponent>
}
